src/client/client/render.js

```diff
@@ -1,377 +1,119 @@
-	stats.update();
-
-	let delta = clock.getDelta();
-
-	uniforms.time.value += delta * 5;
-
-	// Render the darn thing
-	renderer.render( scene, camera );
-}
-
-function spawnMarbleMesh(tags) {
-	let size = tags.size;
-	let color = tags.color;
-	let name = tags.name;
-	let useFancy = tags.useFancy;
-
-	let fancyMaterial = new THREE.ShaderMaterial( {
-
-		uniforms: uniforms,
-		vertexShader: document.getElementById("vertexShader").textContent,
-		fragmentShader: document.getElementById("fragmentShader").textContent
-
-	} );
-
-	let sphereGeometry = new THREE.SphereBufferGeometry(size, 9, 9);
-	/* let sphereGeometry = new THREE.BoxGeometry(.2,.2,.2); */
-	let materialColor = new THREE.Color(color);
-	let sphereMaterial = new THREE.MeshStandardMaterial({ color: materialColor });
-	let sphereMesh = new THREE.Mesh(sphereGeometry, (useFancy ? fancyMaterial : sphereMaterial));
-
-	// Shadows
-	sphereMesh.castShadow = config.graphics.castShadow.marbles;
-	sphereMesh.receiveShadow = config.graphics.receiveShadow.marbles;
-
-	// Add name sprite
-	let nameSprite = makeTextSprite(name);
-
-	// Add to collection
-	marbleMeshes.push(sphereMesh);
-
-	// Add objects to the scene
-	scene.add(marbleMeshes[marbleMeshes.length - 1]);
-	scene.add(nameSprite);
-	marbleMeshes[marbleMeshes.length - 1].add(nameSprite);
-}
-
-//
-
-function makeTextSprite( message )
-{
-	let fontface = "Courier New";
-	let fontsize = 24;
-
-	let canvas = document.createElement("canvas");
-	canvas.width = 256;
-	canvas.height = 64;
-	let context = canvas.getContext("2d");
-	context.font = `Bold ${fontsize}px ${fontface}`;
-	context.textAlign = "center";
-
-	// get size data (height depends only on font size)
-	// let metrics = context.measureText( message );
-	// let textWidth = metrics.width;
-
-	// text color
-	context.fillStyle = "#ffffff";
-
-	context.fillText(message, 128, fontsize);
+	// Load models
+	let modelPromises = {};
+	for (let modelName in this.data.models) {
+		modelPromises[modelName] = new Promise((resolve, reject) => {
+			try {
+				_GLTFLoader.parse(this.data.models[modelName].file, null,
+					function(model) {
+						resolve(model.scene);
+					}, function(error) {
+						console.warn(`Unable to load model (${modelName})`, error);
+						reject("error");
+					}
+				);
+			}
+			catch (error) {
+				// Invalid JSON/GLTF files may end up here
+				console.warn(`Unable to load model (${modelName})`, error);
+				reject("error");
+			}
+		});
+	}
 
-	// canvas contents will be used for a texture
-	let texture = new THREE.Texture(canvas);
-	texture.needsUpdate = true;
-	let spriteMaterial = new THREE.SpriteMaterial({map: texture});
-	let sprite = new THREE.Sprite( spriteMaterial );
-	sprite.scale.set(4, 1, 1.0);
-	return sprite;
-}
+	// Load prefabs
+	let prefabPromises = {};
+	return Promise.all(Object.values(modelPromises)).then(() => {
+		for (let prefabUuid in this.data.prefabs) {
+			prefabPromises[prefabUuid] = new Promise((resolve) => {
```

Comment:
I think they're necessary, the docs do make it sound as if GLTFLoader.parse is rather asynchronous https://threejs.org/docs/#examples/loaders/GLTFLoader.parse

Reply:
I meant for the code after the loading, sorry if that wasn't clear. It's good it's wrapped in the `.then`, but I don't think there's a need to create new `Promise`s for cloning the loaded scene and setting the transforms.

Reply:
Ah yes, this was actually due to me first having the idea to not actually wait for all models to load before starting to parse the prefabs, but only those required for those prefabs. In this situation, the asynchronous nature of the models would propagate to the prefabs and world objects.

Should I try to implement it that way, save it for later, or just get rid of the promises here and not worry about the minimal performance gain?

Reply:
I like the idea (and seeing the map gradually load in like that would look cool too), but it's probably not worth doing right now. I'd remove the additional promises since they won't be needed.